# Let's prep our data

### Get the data

- Download to your computer <a href="https://drive.google.com/file/d/1-798fq909hN3x3UxBYUDZrkEaPTliyvJ/view?usp=sharing"> this 256MB file</a> that contains NYPD summons data from 2015-2025s.


In [23]:
## import libraries
import pandas as pd


In [24]:
# import data
nypd_summons = pd.read_csv("summons-15-25.csv")
nypd_summons.info()

<class 'pandas.DataFrame'>
RangeIndex: 1384334 entries, 0 to 1384333
Data columns (total 18 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   SUMMONS_KEY            1384334 non-null  int64  
 1   SUMMONS_DATE           1384334 non-null  str    
 2   OFFENSE_DESCRIPTION    1384316 non-null  str    
 3   LAW_SECTION_NUMBER     1348831 non-null  str    
 4   LAW_DESCRIPTION        1309406 non-null  str    
 5   SUMMONS_CATEGORY_TYPE  1309321 non-null  str    
 6   AGE_GROUP              1354372 non-null  str    
 7   SEX                    1284382 non-null  str    
 8   RACE                   964769 non-null   str    
 9   JURISDICTION_CODE      1384334 non-null  int64  
 10  BORO                   1372084 non-null  str    
 11  PRECINCT_OF_OCCUR      1384334 non-null  int64  
 12  X_COORDINATE_CD        1383871 non-null  float64
 13  Y_COORDINATE_CD        1383871 non-null  float64
 14  Latitude               138387

## Doesn't AI make learning coding obsolete?

Yes, it's excellent at data analysis and coding. 

But here are a few problems you might run into:

- How will you investigate AI-related topics?
- Hitting Limits – even with paid accounts.
- Protecting sources.
- Not infallibable.

At Bloomberg, the rule is **you are responsible** for every number, analysis and finding you put into your pieces.

So be prepared to understand and explain each of your decisions, including AI's guidance.

## Traditonal coding v. Vibe Coding


**Traditional coding**
- structured approach to learning syntax and rules first.

**Vibe coding**
- learning by experimenting, observing results, analyzing and iterating – with AI's assistance – rather than memorizing syntax.

- still need to learn coding because you must understand what the code is doing, recognizing mistakes and incorrect interpretations of your prompts.

So you need to retain control of your editorial decisions and process. AI can help write code, but understanding coding helps you trust — or question — the answers.


What are some things - terms, processes etc – that we learned so far?
- Python is good at math
- variables are characters in your data.
- PEMDAS
- f-strings
- built in functions
    - like what? len, type, print, abs, round
- importing and reading data
- understanding dataframes v. series
- getting an overview of your data

# Interviewing Data

If AI can write the code, what questions might you ask based on your domain/beat knowledge?

Let's step back to see how we <a href="https://docs.google.com/presentation/d/1oA50DVvcPT2IWDAHTXogD5lWsWk1TVq_sp5iIS1qQv8/edit?usp=sharing">"interview"</a> data.

## Interviewing Data with AI



In [25]:
## Get a sample. 
nypd_summons.sample(10)

,SUMMONS_KEY,SUMMONS_DATE,OFFENSE_DESCRIPTION,LAW_SECTION_NUMBER,LAW_DESCRIPTION,SUMMONS_CATEGORY_TYPE,AGE_GROUP,SEX,RACE,JURISDICTION_CODE,BORO,PRECINCT_OF_OCCUR,X_COORDINATE_CD,Y_COORDINATE_CD,Latitude,Longitude,Lon_Lat,YEAR
45520,289616699,2024-07-04,OBSTRUCTS VEHICULAR OR PEDESTRIAN TRAFFIC,240.20(5),(null),(null),65+,M,ASIAN / PACIFIC ISLANDER,0,MANHATTAN,18,988392.0,215646.0,40.758574,-73.985049,POINT (-73.98504893970802 40.75857392739822),2024
51153,288512581,2024-06-14,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,UNKNOWN,(null),(null),0,BRONX,40,1003714.0,235850.0,40.814010,-73.929683,POINT (-73.929683 40.81401),2024
897375,156340339,2016-08-30,URINATING IN PUBLIC,16-1186,Administrative Code,URINATING,25-44,M,BLACK HISPANIC,0,QUEENS,114,1007977.0,217884.0,40.764694,-73.914348,POINT (-73.91434801399998 40.76469406300004),2016
715101,165721885,2017-06-10,POSS. OF KNIFE,10-133,Administrative Code,KNIFE,65+,M,BLACK,1,MANHATTAN,34,1006041.0,255411.0,40.867700,-73.921215,POINT (-73.92121504699998 40.86770011000005),2017
718390,165608676,2017-06-05,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,25-44,M,BLACK,2,MANHATTAN,7,988428.0,200322.0,40.716522,-73.984930,POINT (-73.98492995199997 40.71652158200004),2017
1037927,151414094,2016-03-15,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,45-64,F,UNKNOWN,0,MANHATTAN,14,986382.0,213900.0,40.753791,-73.992306,POINT (-73.99230628299993 40.75379058100003),2016
7550,296990989,2024-11-23,INSPECTION; PEDICAB REGISTRATION PLATE,20-255,(null),(null),18-24,M,WHITE,0,MANHATTAN,14,987889.0,214095.0,40.754317,-73.986865,POINT (-73.98686541972873 40.754317049095675),2024
756013,163852866,2017-04-14,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,45-64,M,BLACK,0,BROOKLYN,88,992297.0,188389.0,40.683766,-73.970987,POINT (-73.97098738499993 40.68376560900003),2017
628024,314137161,2025-10-08,RED LIGHT (DISOBEY),1111,VTL,VTL,25-44,M,WHITE HISPANIC,0,BROOKLYN,84,987759.0,191193.0,40.691457,-73.987347,POINT (-73.98734658 40.69145665),2025
1317299,142078516,2015-04-11,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,45-64,M,UNKNOWN,0,BROOKLYN,73,1010122.0,187312.0,40.680775,-73.906722,POINT (-73.90672217999997 40.680775340000025),2015


### What are some questions you could ask this data?

1. Which borough had the most summons?
2. How many times did the top 10 most common offenses occur?
3. What are the total number of summons broken down by year?
4. What were some of the most unusual summons? (how do you define most unusual)
5. What is the count of summons by gender?
6. How many summons were issued by race?
7. Which borough had the fewest offenses per 100K people?

### Upload data to AI - ask it questions.


#### What happened?

What do we do on deadline when the dataset is too big?

Even if AI ingests a large dataset, it responds by analyzing a fraction of the data.

The importance of <a href="https://docs.google.com/presentation/d/1BKn8iUq9wy3KgrcfbfpJN6i3cnB8dNlnbAQlrbUSGJ0/edit?usp=sharing">Reproducible Data Journalism</a>

## And now let's interview data using Pandas



We are soon going to blend using AI and coding, but first it's important to understand some Pandas fundamentals.

#### Finally **Vibe Coding**!

##### Template for Vibe Coding

1. Give AI a sample of your data (like a 100 rows)
2. Ask it for the Pandas code to answer a question, **but always tell it to comment the code**!
3. Leave a comment in Markdown with your prompt
4. Test it out in a code cell.
5. If the code throws an error, copy the entire code and give it AI.
6. Remember to add this updated prompt to your markdown.
7. Test out the the updated code in same cell.
8. Create a markdown cell in which you use f-strings to leave your finding

#### 1. Give AI a sample of your data (like a 100 rows) 

In [26]:
## see sample data
summons_sample = nypd_summons.sample(100)
summons_sample

,SUMMONS_KEY,SUMMONS_DATE,OFFENSE_DESCRIPTION,LAW_SECTION_NUMBER,LAW_DESCRIPTION,SUMMONS_CATEGORY_TYPE,AGE_GROUP,SEX,RACE,JURISDICTION_CODE,BORO,PRECINCT_OF_OCCUR,X_COORDINATE_CD,Y_COORDINATE_CD,Latitude,Longitude,Lon_Lat,YEAR
734230,164782595,2017-05-14,DISORDERLY CONDUCT,240.2,Penal Law,DISORDERLY CONDUCT,18-24,M,WHITE,0,QUEENS,108,1002924.0,210628.0,40.744790,-73.932609,POINT (-73.93260926599999 40.74479030600002),2017
1380379,140578228,2015-01-15,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,NaN,D,UNKNOWN,0,BRONX,48,1012605.0,246785.0,40.844006,-73.897519,POINT (-73.89751909499995 40.84400568400008),2015
470583,187296511,2018-09-03,ALCOHOLIC BEVERAGE IN PUBLIC,10-125(2B),NaN,NaN,45-64,M,BLACK,0,BRONX,44,1007027.0,245405.0,40.840234,-73.917684,POINT (-73.91768411 40.84023414),2018
992781,153091313,2016-05-14,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,25-44,M,UNKNOWN,0,BRONX,46,1009381.0,248290.0,40.848146,-73.909166,POINT (-73.90916582199996 40.84814622900007),2016
751870,164022674,2017-04-20,CONSUMPTION OF ALCOHOL,10-125,Administrative Code,ALCOHOL,45-64,M,WHITE,0,STATEN ISLAND,121,939961.0,166911.0,40.624706,-74.159548,POINT (-74.15954752499994 40.62470612700002),2017
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
126018,274354701,2023-09-09,POSS. OF KNIFE,10-133,Administrative Code,KNIFE,18-24,F,WHITE HISPANIC,0,BROOKLYN,73,1007711.0,183018.0,40.668988,-73.915428,POINT (-73.91542835682617 40.66898774658204),2023
433824,193880872,2019-02-22,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,UNKNOWN,NaN,NaN,0,STATEN ISLAND,122,954827.0,148519.0,40.574286,-74.105914,POINT (-74.10591440499996 40.57428568800003),2019
747288,164222561,2017-04-26,POSS. OF KNIFE,10-133,Administrative Code,KNIFE,18-24,M,WHITE HISPANIC,0,MANHATTAN,20,991296.0,225555.0,40.785778,-73.974558,POINT (-73.97455764699998 40.785777977000066),2017
1266597,143518754,2015-06-06,BIKE IN PARK,1-05I,Park Regs,BIKE,25-44,M,NaN,0,NaN,60,990922.0,149184.0,40.576157,-73.975984,POINT (-73.97598379999994 40.576157300000034),2015


In [27]:
# Summon the Head 
nypd_summons.head()

,SUMMONS_KEY,SUMMONS_DATE,OFFENSE_DESCRIPTION,LAW_SECTION_NUMBER,LAW_DESCRIPTION,SUMMONS_CATEGORY_TYPE,AGE_GROUP,SEX,RACE,JURISDICTION_CODE,BORO,PRECINCT_OF_OCCUR,X_COORDINATE_CD,Y_COORDINATE_CD,Latitude,Longitude,Lon_Lat,YEAR
0,298767983,2024-12-31,DISORDERLY CONDUCT,240.2,Penal Law,DISORDERLY CONDUCT,45-64,M,WHITE HISPANIC,0,MANHATTAN,14,985927.0,213077.0,40.751523,-73.993947,POINT (-73.99394730218468 40.75152348693427),2024
1,298704187,2024-12-31,IMPROPER PLATES,402,VTL,VTL,18-24,M,ASIAN / PACIFIC ISLANDER,0,BRONX,49,1020823.0,249562.0,40.851589,-73.867801,POINT (-73.86780058180116 40.85158922344624),2024
2,298787872,2024-12-31,TRESPASS,140.05,Penal Law,TRESPASS,25-44,F,BLACK HISPANIC,2,MANHATTAN,32,998922.0,235604.0,40.813352,-73.946996,POINT (-73.946996 40.813352),2024
3,298724988,2024-12-31,FEDERAL MOTOR VEH. SAFETY REG,CFR 49,NYS Transportation,NYS TRANS,UNKNOWN,(null),(null),0,BROOKLYN,94,1001547.0,203048.0,40.723980,-73.937597,POINT (-73.93759674847603 40.72397961334044),2024
4,298704460,2024-12-31,POSSESSION OF ALCOHOL BY A MINOR,65-C,(null),(null),18-24,M,WHITE,0,MANHATTAN,6,984439.0,204595.0,40.728254,-73.999318,POINT (-73.999318 40.728254),2024


In [28]:
## export
summons_sample.to_csv("sample_summons_100.csv", index = False, encoding = "UTF8")

### Explanation of the export code

`summons_sample` is the dataframe we want to export

`.to_csv()` saves our dataframe to a .csv file. You are baseically exporting your python data out into a format that Excel, Google Sheets can read.

`index=False`
By default, on export pandas includes the DataFrame's index – the row numbers (0, 1, 2, 3...). We NEVER
need it 99.99 percent of the time. `index=False` simply tells pandas to leave that column out on export.

`encoding="utf-8"`
Tells Python which character encoding to use when writing the file. `UTF-8` is the most universal standard — it handles English plus accented characters, symbols, non-Latin scripts, etc. without garbling them. If your data has names like "García" or "Müller" and you skip this, you might get weird characters or errors depending on your system.

Recall some questions we want to ask:

1. Which borough had the most summons?
2. How many times did the top 10 most common offenses occur?
3. What are the total number of summons broken down by year?
4. What were some of the most unusual summons? (how do you define most unusual)
5. What is the count of summons by gender?
6. How many summons were issued by race?
7. Which borough had the fewest offenses per 100K people?

## Who wants to share their answers?

# Prompt for all questions #
For all the following questions, use Pandas 2.5 or newer.
Always provide comments for each line of code. 




### Prompt: 

Provide the pandas code to calculate which boroughs had the most summons (in descending order).


In [29]:
# 1. Which borough had the most summons? 

# Replace string '(null)' placeholders with real NaN so they're excluded cleanly
nypd_summons['BORO'] = nypd_summons['BORO'].replace('(null)', pd.NA)

# Count summons per borough, sort descending, and drop any missing borough values
borough_counts = (
    nypd_summons['BORO']
    .value_counts()              # Count occurrences of each borough; sorts descending by default
    .reset_index()               # Convert the Series into a two-column DataFrame
    .rename(columns={            # Rename columns from pandas defaults to readable names
        'BORO': 'borough',
        'count': 'summons_count'
    })
)

borough_counts


,borough,summons_count
0,BROOKLYN,427679
1,BRONX,312327
2,MANHATTAN,275884
3,QUEENS,246314
4,NEW YORK,63056
5,STATEN ISLAND,46824


In [30]:
# 2. What were the top 10 most common offenses?

## Input: In the summons-15-25.csv dataset: What were the top 10 most common offenses? Return the answer in a dataframe written in Python.

## Output:

top_10_offenses = pd.DataFrame(
    {
        "OFFENSE_DESCRIPTION": [
            "CONSUMPTION OF ALCOHOL",
            "FEDERAL MOTOR VEH. SAFETY REG",
            "MARIJUANA, POSSESSION OF",
            "URINATING IN PUBLIC",
            "RECKLESS DRIVING",
            "ALCOHOLIC BEVERAGE IN PUBLIC",
            "FAILURE TO COMPLY WITH SIGN",
            "TRESPASS",
            "OTHER ADMIN CODE",
            "DISORDERLY CONDUCT",
        ],
        "COUNT": [248650, 158420, 84910, 50340, 44691, 43789, 42810, 38595, 31851, 20743],
    }
)

top_10_offenses

,OFFENSE_DESCRIPTION,COUNT
0,CONSUMPTION OF ALCOHOL,248650
1,FEDERAL MOTOR VEH. SAFETY REG,158420
2,"MARIJUANA, POSSESSION OF",84910
3,URINATING IN PUBLIC,50340
4,RECKLESS DRIVING,44691
5,ALCOHOLIC BEVERAGE IN PUBLIC,43789
6,FAILURE TO COMPLY WITH SIGN,42810
7,TRESPASS,38595
8,OTHER ADMIN CODE,31851
9,DISORDERLY CONDUCT,20743


In [31]:
# 3. What are the total number of summons broken down by year?

## Input: In the summons-15-25.csv dataset: What are the total number of summons broken down by year? Return the answer in a dataframe written in Python.

## Output:

summons_by_year = pd.DataFrame(
    {
        "YEAR": [2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024, 2025],
        "SUMMONS_COUNT": [297748, 267950, 165093, 89895, 85899, 55589, 45717, 58059, 106529, 94542, 117313],
    }
)

summons_by_year

,YEAR,SUMMONS_COUNT
0,2015,297748
1,2016,267950
2,2017,165093
3,2018,89895
4,2019,85899
5,2020,55589
6,2021,45717
7,2022,58059
8,2023,106529
9,2024,94542


In [32]:
# 4. What were some of the most unusual summons?

## Input: In the summons-15-25.csv dataset: What offenses were repeated 5 or fewer times within this dataset? Return a dataframe written in Python containing each offense and the number of times it occurred in this dataset.

## Output:

path = "/Users/jackwalker/dataProject/2026-03-17/summons-15-25.csv"

offenses_5_or_fewer = (
    pd.read_csv(path, usecols=["OFFENSE_DESCRIPTION"])["OFFENSE_DESCRIPTION"]
    .fillna("UNKNOWN")
    .value_counts()
    .rename_axis("OFFENSE_DESCRIPTION")
    .reset_index(name="COUNT")
    .query("COUNT <= 5")
    .sort_values(["COUNT", "OFFENSE_DESCRIPTION"], ascending=[True, True])
    .reset_index(drop=True)
)

offenses_5_or_fewer

,OFFENSE_DESCRIPTION,COUNT
0,ACCIDENT WITH INJURY/DAMAGE; FAIL TO STOP AND ...,1
1,ALLOW CONSUMPTION BETWEEN 4:00 AM - 12 NOON (S...,1
2,ANIMAL/DOG IN PARK BUILDING/BATH FACILITY,1
3,AUDIO EQUIPMENT; FAIL TO OBLIGE PASSENGER REGA...,1
4,BICYCLE NO BELL,1
...,...,...
150,SAFETY ZONE DIRECTION,5
151,"SCHOOL BUS, PASSING STOPPED",5
152,SELL/LOAN/LEASE/TRANSFER,5
153,UNLICENSED VEHICLE; KNOWINGLY OPERATE FOR HIRE,5


In [33]:
# 5. What is the count of summons by gender?

## Input: In the summons-15-25.csv dataset: What was the gender breakdown of offenses? Return the answer in a dataframe written in Python.

## Output:

gender_breakdown = pd.DataFrame(
    {
        "SEX": ["M", "UNKNOWN", "F", "D", "U"],
        "OFFENSE_COUNT": [1013591, 205544, 130370, 29961, 4868],
    }
)

gender_breakdown

,SEX,OFFENSE_COUNT
0,M,1013591
1,UNKNOWN,205544
2,F,130370
3,D,29961
4,U,4868


In [34]:
# 6. How many summons were issued by race?


## Input: In the summons-15-25.csv dataset: What was the racial breakdown of offenses? Return the answer in a dataframe written in Python.

## Output:

racial_breakdown = pd.DataFrame(
    {
        "RACE": [
            "UNKNOWN",
            "BLACK",
            "WHITE HISPANIC",
            "WHITE",
            "BLACK HISPANIC",
            "ASIAN / PACIFIC ISLANDER",
            "AMERICAN INDIAN/ALASKAN NATIVE",
            "OTHER",
        ],
        "OFFENSE_COUNT": [659810, 316676, 192409, 87448, 70375, 49862, 5735, 2019],
    }
)

racial_breakdown

,RACE,OFFENSE_COUNT
0,UNKNOWN,659810
1,BLACK,316676
2,WHITE HISPANIC,192409
3,WHITE,87448
4,BLACK HISPANIC,70375
5,ASIAN / PACIFIC ISLANDER,49862
6,AMERICAN INDIAN/ALASKAN NATIVE,5735
7,OTHER,2019


# 2020 Census population
```python
populations = {
    "BROOKLYN": 2736074,
    "QUEENS": 2405464,
    "MANHATTAN": 1694251,
    "BRONX": 1472654,
    "STATEN ISLAND": 495747
}
```

In [35]:
#7. Which borough had the fewest offenses per 100K people?

# 2020 Census population
populations = {
    "BROOKLYN": 2736074,
    "QUEENS": 2405464,
    "MANHATTAN": 1694251,
    "BRONX": 1472654,
    "STATEN ISLAND": 495747
}

In [36]:
## Input:Given the following data, what borough had the fewest offenses per 100K people?

##populations = {
##"BROOKLYN": 2736074,
##"QUEENS": 2405464,
##"MANHATTAN": 1694251,
##"BRONX": 1472654,
##"STATEN ISLAND": 495747
##}

## Output: 

populations = {
    "BROOKLYN": 2736074,
    "QUEENS": 2405464,
    "MANHATTAN": 1694251,
    "BRONX": 1472654,
    "STATEN ISLAND": 495747
}

offenses_per_100k = pd.DataFrame(
    {
        "BOROUGH": ["STATEN ISLAND", "QUEENS", "BROOKLYN", "MANHATTAN", "BRONX"],
        "OFFENSE_COUNT": [46824, 246314, 427679, 275884, 312327],
        "POPULATION": [495747, 2405464, 2736074, 1694251, 1472654],
        "OFFENSES_PER_100K": [9445.140364, 10239.770789, 15631.119626, 16283.537681, 21208.444074],
    }
).sort_values("OFFENSES_PER_100K").reset_index(drop=True)

offenses_per_100k

,BOROUGH,OFFENSE_COUNT,POPULATION,OFFENSES_PER_100K
0,STATEN ISLAND,46824,495747,9445.140364
1,QUEENS,246314,2405464,10239.770789
2,BROOKLYN,427679,2736074,15631.119626
3,MANHATTAN,275884,1694251,16283.537681
4,BRONX,312327,1472654,21208.444074


Therefore, **the Bronx** had the most offenses per 100K people, with 21,208.4 offenses per 100K people.

## Which AI agent did I use?

I ran this dataset through Visual Studio Code and used **GitHub's Copilot** to analyze it.